### PR TITLE
Added option to replace dots in keys with underscores

### PR DESCRIPTION
**Pull request by MarcusNoble**
*Issue #, if available:* N/A

*Description of changes:*
Add an option to replace dots (`.`) within keys with underscores (`_`) to help avoid type clashes in Elasticsearch and similar.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: kinesis/kinesis.go [hunk 1]
```diff
@@ -185,6 +189,32 @@ func (outputPlugin *OutputPlugin) Flush() error {
...
+		switch kt := k.(type) {
+		case string:
+			k = strings.ReplaceAll(kt, ".", "_")
+		}
+
+		switch vt := v.(type) {
+		case map[interface{}]interface{}:
+			v = replaceDots(vt)
+			// Ensure duplicated keys are merged rather than replaced
+			if _, ok := newObj[k]; ok {
```
Comment (PettitWesley) [inline kinesis/kinesis.go]: checking for duplicated keys needs to happen in the first call to `replaceDots`, not just in the recursive case

File: kinesis/kinesis.go [hunk 2]
```diff
@@ -185,6 +189,32 @@ func (outputPlugin *OutputPlugin) Flush() error {
 	return outputPlugin.sendCurrentBatch()
 }
 
+func replaceDots(obj map[interface{}]interface{}) map[interface{}]interface{} {
+	newObj := make(map[interface{}]interface{})
```
Comment (PettitWesley) [inline kinesis/kinesis.go]: I'd like this code to work without requiring an additional map allocation... that's going to add a lot overhead.

File: kinesis/kinesis.go (hunk 1 above)
Comment (PettitWesley) [reply kinesis/kinesis.go]: Interestingly, the C code for the ES version of this option is very simple and does not check for duplicates:

https://github.com/fluent/fluent-bit/blob/master/plugins/out_es/es.c#L90

Comment (PettitWesley) [reply kinesis/kinesis.go]: Checking for duplicated keys is safer; I like having that check. However, I think it should be able to be done in place without allocating a new map.

I believe this should be safe in Go: https://stackoverflow.com/questions/23229975/is-it-safe-to-remove-selected-keys-from-map-within-a-range-loop

**Comment by MarcusNoble** [comment]
Thanks for your review @PettitWesley. 

I've removed the re-assignment so now it just does a mutation in place. 

I've also removed the deduplication code as that needs some more thought. I'm going to open an issue to discuss it and provide the scenario that I'm hitting a problem.

**Comment by PettitWesley** [comment]
@MarcusNoble I lean towards saying this should be a feature in Fluent Bit core; see here: https://github.com/fluent/fluent-bit/issues/1305#issuecomment-585993039

**Comment by nsvijay04b1** [comment]
i am one of those who are waiting for this merge so i can use it :-)

**Comment by nsvijay04b1** [comment]
hi .. Any progress please? awaiting fix as i have the same problem.

**Comment by nsvijay04b1** [comment]
Release v2.7.0  also missed this feature.   Hope this will be part of Release v2.8.0

**Comment by hossain-rayhan** [comment]
Hi @nsvijay04b1 , as @PettitWesley  mentioned, this might a good option to put into the `fluent-bit` core. So, I am not sure if we are planning to implement it separately for each of our plugins. 

Check this for update: https://github.com/fluent/fluent-bit/issues/1305#issuecomment-585993039

**Comment by PettitWesley** [comment]
We're now thinking of working on this and including it in an upcoming release

**Comment by PettitWesley** [comment]
@MarcusNoble @nsvijay04b1 We have take this back up in #79. Marcus's original commits are preserved.

Apologies for taking so long to deliberate on this.

**Pull request closed without merging**